src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs

```diff
@@ -0,0 +1,114 @@
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    internal class UpholdBuildConventionRule : IPackageRule
+    {
+        private static ManagedCodeConventions ManagedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+        private string _propsExtension = ManagedCodeConventions.Properties["msbuild"].FileExtensions[1];
+        private string _targetsExtension = ManagedCodeConventions.Properties["msbuild"].FileExtensions[0];
+
+        public string MessageFormat => AnalysisResources.BuildConventionIsViolatedWarning;
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var files = builder.GetFiles().Where(t => t.StartsWith(PackagingConstants.Folders.Build));
+            var packageId = builder.NuspecReader.GetId();
+            var (conventionViolatorsProps, conventionViolatorsTargets) = IdentifyViolators(files, packageId);
+            return GenerateWarnings(conventionViolatorsProps, conventionViolatorsTargets);
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(IDictionary<string, string> conventionViolatorsProps, IDictionary<string, string> conventionViolatorsTargets)
+        {
+            var issues = new List<PackagingLogMessage>();
+            foreach (var props in conventionViolatorsProps)
+            {
+                var warning = string.Format(MessageFormat, _propsExtension, props.Key, props.Value);
+                issues.Add(PackagingLogMessage.CreateWarning(warning, NuGetLogCode.NU5129));
+            }
+
+            foreach (var targets in conventionViolatorsTargets)
+            {
+                var warning = string.Format(MessageFormat, _targetsExtension, targets.Key, targets.Value);
+                issues.Add(PackagingLogMessage.CreateWarning(warning, NuGetLogCode.NU5129));
+            }
+            return issues;
+        }
+
+        internal (IDictionary<string, string>, IDictionary<string, string>) IdentifyViolators(IEnumerable<string> files, string packageId)
+        {
+            var collection = new ContentItemCollection();
+            collection.Load(files.Where(t => PathUtility.GetPathWithDirectorySeparator(t).Count(m => m == Path.DirectorySeparatorChar) > 1));
+
+            var buildItems = ContentExtractor.GetContentForPattern(collection, ManagedCodeConventions.Patterns.MSBuildFiles);
+            var tfms = ContentExtractor.GetGroupFrameworks(buildItems).Select(m => m.GetShortFolderName());
+
+            var conventionViolatorsProps = new Dictionary<string, string>();
+            var conventionViolatorsTargets = new Dictionary<string, string>();
+
+            var correctProps = packageId + _propsExtension;
```

Comment:
naming: correctPropsPattern & correctTargetsPattern. 

The below was tricky to follow as too many var names were similar.